hottbox/algorithms/decomposition/cpd.py

```diff
@@ -406,3 +421,193 @@ def _init_fmat(self, tensor, rank):
+
+    References
+    ----------
+
+    ..  [1] Kiers, H., ten Berge, J. and Bro, R. (1999). PARAFAC2 - Part I.
+        A direct fitting algorithm for the PARAFAC2 model. Journal of Chemometrics,
+        13(3-4), pp.275-294.
+    """
+    # TODO: change init use requiring a change in TensorCPD
+    def __init__(self, max_iter=50, epsilon=10e-3, tol=10e-5,
+                 random_state=None, verbose=False) -> None:
+        super(Parafac2, self).__init__(init='random',
+                                       max_iter=max_iter,
+                                       epsilon=epsilon,
+                                       tol=tol,
+                                       random_state=random_state,
+                                       verbose=verbose)
+        self.cost = []
+
+    def copy(self):
+        """ Copy of the CPD algorithm as a new object """
+        new_object = super(Parafac2, self).copy()
+        new_object.cost = []
+        return new_object
+
+    @property
+    def name(self):
+        """ Name of the decomposition
+
+        Returns
+        -------
+        decomposition_name : str
+        """
+        decomposition_name = super(Parafac2, self).name
+        return decomposition_name
+
+    # TODO: Parameters differ to base class decomposed - fix
+    def decompose(self, tenL, rank):
+        """ Performs Direct fitting using ALS on the ``tensor`` with respect to the specified ``rank``
+
+        Parameters
+        ----------
+        tenL : List(np.ndarray)
+            List of ``Tensor`` data to be decomposed
+        rank : tuple
+            Desired Kruskal rank for the given ``tensor``. Should contain only one value.
+            If it is greater then any of dimensions then random initialisation is used
+
+        Returns
+        -------
+        fmat_u, fmat_s, fmat_V, reconstructed : Tuple(np.ndarray)
+            fmat_u,fmat_s,fmat_V are PARAFAC2 representation of list of tensors
+            reconstructed is the reconstruction of the original tensor directly using fmat_u, fmat_s, fmat_V
+
+        Notes
+        -----
+        khatri-rao product should be of matrices in reversed order. But this will duplicate original data (e.g. images)
+        Probably this has something to do with data ordering in Python and how it relates to kr product
+        """
+        if not isinstance(tenL, list):
+            raise TypeError("Parameter `tenL` should be a list of np.ndarray objects!")
+        if not all(isinstance(m, np.ndarray) for m in tenL):
+            raise TypeError("Parameter `tenL` should be a list of np.ndarray objects!")
+        if not isinstance(rank, tuple):
+            raise TypeError("Parameter `rank` should be passed as a tuple!")
+        if len(rank) != 1:
+            raise ValueError("Parameter `rank` should be tuple with only one value!")
+
+        sz = np.array([t.shape for t in tenL])
+        _m = list(sz[:, 1])
+        if _m[1:] != _m[:-1]:
```

Comment:
This check does the job correctly, but it took a little bit of time to understand how it works)
Something like this is more readable 
```python
common_size = tenL[0].shape[1]
if not all([t.shape[1] == common_size for t in tenL]):
    raise ValueError
```

Another option, but I like it less.
```python
if np.unique([t.shape[1] for t in tenL]).size =! 1:
    raise ValueError
```

Reply:
I did this because I had to get all the shapes anyway - this way I only iterate through the list once. I am not sure on this one?